cloudinit/ssh_util.py

```diff
@@ -240,25 +240,47 @@ def render_authorizedkeysfile_paths(value, homedir, username):
         value = "%h/.ssh/authorized_keys"
     paths = value.split()
     rendered = []
+    user_authorized = default_auth
     for path in paths:
-        for macro, field in macros:
-            path = path.replace(macro, field)
+        for i in range(len(macros)):
+            macro = macros[i][0]
+            field = macros[i][1]
+            path_replaced = path.replace(macro, field)
+            # if %h or %u in path, it means it is a local key
+            if i < 2 and macro in path:
+                user_authorized = path_replaced
+            path = path_replaced
         if not path.startswith("/"):
             path = os.path.join(homedir, path)
         rendered.append(path)
-    return rendered
+    return rendered, user_authorized
 
 
 def extract_authorized_keys(username, sshd_cfg_file=DEF_SSHD_CFG):
     (ssh_dir, pw_ent) = users_ssh_info(username)
     default_authorizedkeys_file = os.path.join(ssh_dir, 'authorized_keys')
+    user_authorizedkey = default_authorizedkeys_file
+    pw_len = len(pw_ent.pw_dir)
     auth_key_fns = []
     with util.SeLinuxGuard(ssh_dir, recursive=True):
         try:
             ssh_cfg = parse_ssh_config_map(sshd_cfg_file)
-            auth_key_fns = render_authorizedkeysfile_paths(
-                ssh_cfg.get("authorizedkeysfile", "%h/.ssh/authorized_keys"),
-                pw_ent.pw_dir, username)
+            auth_key_fns, user_authorizedkey = \
+                render_authorizedkeysfile_paths(
+                    ssh_cfg.get("authorizedkeysfile",
+                                "%h/.ssh/authorized_keys"),
+                    pw_ent.pw_dir, username, user_authorizedkey)
+            # check if one of the keys is the user's one
+            for auth_key_fn in auth_key_fns:
+                if auth_key_fn[:pw_len] == pw_ent.pw_dir:
```

Comment:
I understand here you're assuming the only possibility for custom `authorized_keys` file is something that's inside the user home dir. But what if it's not? Like `/var/cloud/%u/keys`?

Reply:
This case is handled in `render_authorizedkeysfile_paths`: if `%u` or %h` is present in the path, it is a valid user-specific file and we update user_authorized. Then, in this case I perform the above check anyways (could theoretically be skipped if user_authorizedkey != default_authorizedkeys_file) , and see if there are matching strings with the user home path. Hope this clarifies it!